ob1k-concurrent/src/test/java/com/outbrain/ob1k/concurrent/ComposableFutureTest.java

```diff
@@ -169,6 +170,35 @@ public void testDelayedFuture() throws Exception {
     assertTrue("thread should have been waiting for at least 500ms (1s delay)", timeTook > 500);
   }
 
+  @Test
+  public void testFlattenSuccess() throws Exception {
+    final ComposableFuture<String> successFuture = fromValue("deep");
+    final ComposableFuture<ComposableFuture<String>> successNestedFuture = fromValue(successFuture);
+    final ComposableFuture<String> successFlatten = flatten(successNestedFuture);
+
+    assertEquals("flatten and original should have same value", successFuture.get(), successFlatten.get());
+  }
+
+  @Test
+  public void testFlattenFailure() throws Exception {
+    final ComposableFuture<String> errorFuture = fromError(new IllegalStateException("errr"));
+    final ComposableFuture<ComposableFuture<String>> errorNestedFuture = fromValue(errorFuture);
+    final ComposableFuture<String> errorFlatten = flatten(errorNestedFuture);
+
+    Throwable cause = getFutureCause(errorFuture);
+    Throwable causeFlatten = getFutureCause(errorFlatten);
+    assertEquals("flatten and original should have same value", cause, causeFlatten);
```

Comment:
no really need in this helper method, I'd just change @Test to expect IllegalStateException and do errorFlatten.get();
end of test

try to keep it simple and clean :)

Reply:
Thanks, will change that.